pkg/volume/csi/csi_block.go

```diff
@@ -0,0 +1,271 @@
+	attachID := getAttachmentName(csiSource.VolumeHandle, csiSource.Driver, nodeName)
+
+	// search for attachment by VolumeAttachment.Spec.Source.PersistentVolumeName
+	attachment, err := m.k8s.StorageV1beta1().VolumeAttachments().Get(attachID, meta.GetOptions{})
+	if err != nil {
+		glog.Error(log("blockMapper.MapDevice failed to get volume attachment [id=%v]: %v", attachID, err))
+		return err
+	}
+
+	if attachment == nil {
+		glog.Error(log("blockMapper.MapDevice unable to find VolumeAttachment [id=%s]", attachID))
+		return errors.New("no existing VolumeAttachment found")
+	}
+	publishVolumeInfo := attachment.Status.AttachmentMetadata
+
+	nodePublishSecrets := map[string]string{}
+	if csiSource.NodePublishSecretRef != nil {
+		nodePublishSecrets, err = getCredentialsFromSecret(m.k8s, csiSource.NodePublishSecretRef)
+		if err != nil {
+			glog.Errorf("blockMapper.MapDevice failed to get NodePublishSecretRef %s/%s: %v",
+				csiSource.NodePublishSecretRef.Namespace, csiSource.NodePublishSecretRef.Name, err)
+			return err
+		}
+	}
+
+	if err := os.MkdirAll(dir, 0750); err != nil {
+		glog.Error(log("blockMapper.MapDevice failed to create dir %#v:  %v", dir, err))
+		return err
+	}
+	glog.V(4).Info(log("blockMapper.MapDevice created NodePublish path [%s]", dir))
+
+	//TODO (vladimirvivien) implement better AccessModes mapping between k8s and CSI
+	accessMode := v1.ReadWriteOnce
+	if m.spec.PersistentVolume.Spec.AccessModes != nil {
+		accessMode = m.spec.PersistentVolume.Spec.AccessModes[0]
+	}
+
+	err = csi.NodePublishVolume(
+		ctx,
+		m.volumeID,
+		m.readOnly,
+		globalMapPath,
+		dir,
+		accessMode,
+		publishVolumeInfo,
+		csiSource.VolumeAttributes,
+		nodePublishSecrets,
+		fsTypeBlockName,
+	)
+
+	if err != nil {
+		glog.Errorf(log("blockMapper.MapDevice failed: %v", err))
+		if err := os.RemoveAll(dir); err != nil {
+			glog.Error(log("blockMapper.MapDevice failed to remove mount dir after a NodePublish() error [%s]: %v", dir, err))
+		}
+		return err
+	}
+
+	return nil
+}
+
+var _ volume.BlockVolumeUnmapper = &csiBlockMapper{}
+
+// TearDownDevice removes traces of the SetUpDevice.
+func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error {
+	glog.V(4).Infof(log("unmapper.TearDownDevice(globalMapPath=%s; devicePath=%s)", globalMapPath, devicePath))
+
+	csi := m.csiClient
+	ctx, cancel := context.WithTimeout(context.Background(), csiTimeout)
+	defer cancel()
+
+	// unmap global device map path
+	if err := csi.NodeUnstageVolume(ctx, m.volumeID, globalMapPath); err != nil {
+		glog.Errorf(log("blockMapper.TearDownDevice failed: %v", err))
+		return err
+	}
+	glog.V(4).Infof(log("blockMapper.TearDownDevice NodeUnstageVolume successfully [%s]", globalMapPath))
+
+	// request to remove pod volume map path also
+	podVolumePath, volumeName := m.GetPodDeviceMapPath()
+	podVolumeMapPath := filepath.Join(podVolumePath, volumeName)
+	if err := csi.NodeUnpublishVolume(ctx, m.volumeID, podVolumeMapPath); err != nil {
```

Comment:
Shouldn't `NodeUnstageVolume` and `NodeUnpublishVolume` be called by `TearDownDevice` and `UnmapDevice` instead of both in the same method?

Reply:
@saad-ali yeah the API probably should have an `UnmapDevice` method.  But It does not.  Similarly, the `MapDevice` was introduced in the API with the refactor PR.  So, maybe later we do that.